### DISCUSSION TITLE
Codesys structures containing arrays in python-opcua

### DISCUSSION THREAD

**Discussion by tkubt** [FreeOpcUa/python-opcua#1315] [Q&A]
Hi, i am trying to read/write from Codesys structures containing arrays (https://help.codesys.com/webapp/_cds_datatype_structure;product=codesys;version=3.5.17.0), where a Wago PFC200 is the server. I can read/write the instLoad of type structLoad containing two DINT-arrays with UaExpert client:
![grafik](https://user-images.githubusercontent.com/36536653/118246669-93a99c00-b4a2-11eb-8025-87a3a0395160.png)

In python-opcua i do:
```python
from opcua import Client
from opcua import ua

client = Client("opc.tcp://xxx.xxx.xxx.xxx:4840/") # ip hidden
client.connect()
root = client.get_root_node()
client.load_type_definitions()
node = client.get_node('ns=4;s=|var|PFC200.Application.GVL.instLoad') # get node via id
recvStruct = node.get_value() 
print(recvStruct)
```
returns 
```python
structLoad(arrdiLoadPred_Size: 3, arrdiLoadPred: 1, arrdiLoadMeas_Size: 0, arrdiLoadMeas: 0)
```
and 
```python
ua.structLoad.ua_types and recvStruct.ua_types
```
both return
```python
[('arrdiLoadPred_Size', 'Int32'),
 ('arrdiLoadPred', 'Int32'),
 ('arrdiLoadMeas_Size', 'Int32'),
 ('arrdiLoadMeas', 'Int32')]
```
The xml-file, which is autogenerated by Codesys/Wago contains:
```xml
    <TypeUserDef name="T_structLoad" size="24" nativesize="24" typeclass="Userdef" pouclass="STRUCTURE" iecname="structLoad">
      <UserDefElement iecname="arrdiLoadPred" type="T_ARRAY__1__3__OF_DINT" byteoffset="0" vartype="VAR" />
      <UserDefElement iecname="arrdiLoadMeas" type="T_ARRAY__1__3__OF_DINT" byteoffset="12" vartype="VAR" />
    </TypeUserDef>
```
Why is arrdiLoadMeas and arrdiLoadMeas_Size zero? Why are the arrays of type Int32? Is there a way to interpret Codesys structures as nested list or similar?

Thank you and best regards
Tobias

**Comment by AndreasHeine** [comment]
i am not sure if i understand the question... have you used the Debugger and some breakpoints to investigate what python object returned on read?

**Reply by tkubt** [reply]
Hi, sorry if i wasn't clear enough. How can i access the values in these two arrays? Here is some output on the Python object recvStruct:

```
type(recvStruct)
structLoad
```
```
print(recvStruct.__dict__)
{'arrdiLoadPred_Size': 3, 'arrdiLoadPred': 1, 'arrdiLoadMeas_Size': 0, 'arrdiLoadMeas': 0}
# arrdiLoadMeas_Size should be 3
```
```
type(recvStruct.arrdiLoadPred) # this should return list or similar, because arrdiLoadPred is an Codesys array.
int 
```
```
print(ua.structLoad.ua_types)
[('arrdiLoadPred_Size', 'Int32'), ('arrdiLoadPred', 'Int32'), ('arrdiLoadMeas_Size', 'Int32'), ('arrdiLoadMeas', 'Int32')]
```

```
print(recvStruct.ua_types)
[('arrdiLoadPred_Size', 'Int32'), ('arrdiLoadPred', 'Int32'), ('arrdiLoadMeas_Size', 'Int32'), ('arrdiLoadMeas', 'Int32')]
```
```
print(recvStruct.arrdiLoadMeas) # in Codesys arrdiLoadMeas[2] is 1
0
```
In Codesys instLoad is an instance of structLoad:
```
instLoad : structLoad;
```
And structLoad is a STRUCT containing two arrays of DINT:
```
TYPE structLoad :
STRUCT
	arrdiLoadPred : ARRAY [1..3] OF DINT;
	arrdiLoadMeas : ARRAY [1..3] OF DINT;
END_STRUCT
END_TYPE
```

**Reply by oroulet** [reply]
port your code to opcua.asyncio using the ysnc api, it should be quick and easy, then see f you stil lhave issues. I do not remember the status of struct support but it is clearly better in asyncua
